### PR TITLE
Support Uploading Excel Files

### PR DESCRIPTION
**Pull request by rajadain**
## Overview

Adds support for uploading Excel files. We support both `.xls` and `.xlsx` files. Only the first sheet of the file is read, and is expected to be in the same format as the CSVs we expect.

Connects #484

## Testing Instructions

* Check out this branch and try to upload an Excel file (now there are two sample files in the Facility Lists directory)
  - [x] Ensure `.xls` upload works
  - [x] Ensure `.xlsx` upload works
* Download the CSV template from the site and open in Excel or Numbers or Open Office
* Add some rows. Save the file as an Excel file.
* Upload the new file.
  - [x] Ensure the upload works

## Checklist

- [x] `fixup!` commits have been squashed
- [x] CI passes after rebase
- [x] CHANGELOG.md updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines

### CODE REVIEW

File: src/django/api/tests.py [hunk 1]
```diff
@@ -55,6 +57,20 @@ def setUp(self):
                        [s.encode() for s in self.test_csv_rows[1:]]),
             content_type='text/csv')
 
+        lists_dir = '/usr/local/src/api/management/commands/facility_lists/'
+        with open(os.path.join(lists_dir, '12.xls'), 'rb') as xls:
+            self.test_file_xls = SimpleUploadedFile(
+                '12.xls',
+                xls.read(),
+                content_type='application/vnd.ms-excel')
```
[the remark this answers is not in the document]
Comment (rajadain) [reply src/django/api/tests.py]: I got these from reading best practice documentation online, but you're right: we should probably mimic what happens when we actually upload from the browser. Will observe and modify accordingly.

File: src/django/api/processing.py [hunk 2]
```diff
@@ -20,6 +21,18 @@
 from api.geocoding import geocode_address
 
 
+def rollbar_error(file, request):
```
[the remark this answers is not in the document]
Comment (rajadain) [reply src/django/api/processing.py]: Currently it _is_ used outside the module:

https://github.com/open-apparel-registry/open-apparel-registry/blob/0a40df5b0239bb013f32be731031f521ac259aeb/src/django/api/views.py#L602-L604

But I'm thinking of a way to move all that in to `processing.py` which would support scoping this.

File: src/django/api/processing.py [hunk 3]
```diff
@@ -33,6 +34,20 @@ def rollbar_error(file, request):
...
+    from defusedxml.common import EntitiesForbidden
+
+    defusedxml.defuse_stdlib()
+
+    try:
+        return xlrd.open_workbook(file_contents=file.read(),
+                                  on_demand=True).sheet_by_index(0)
+    except EntitiesForbidden:
+        rollbar_error(file, request)
+        raise ValidationError('Error opening Excel file')
```
[the remark this answers is not in the document]
Comment (rajadain) [reply src/django/api/processing.py]: Fixed in 81b0900

File: src/django/api/processing.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rajadain) [reply src/django/api/processing.py]: Fixed in c8c4290

File: src/django/api/views.py [hunk 4]
```diff
@@ -589,6 +587,38 @@ def _validate_header(self, header):
...
+                        for idx in range(1, sheet.nrows)]
+            except Exception:
+                rollbar_error(file, request)
+                raise ValidationError('Error parsing Excel file')
+        elif ext == '.csv':
+            try:
+                header = file.readline().decode(encoding='utf-8-sig').rstrip()
+                self._validate_header(header)
+
+                rows = [line.decode(encoding='utf-8-sig').rstrip()
```
[the remark this answers is not in the document]
Comment (rajadain) [reply src/django/api/views.py]: Fixed in c8c4290

File: src/django/api/tests.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rajadain) [reply src/django/api/tests.py]: Inspecting uploads from three browsers, these are the mime types I'm seeing:

```
Chrome:
.xls = application/octet-stream
.xlsx = application/vnd.openxmlformats-officedocument.spreadsheetml.sheet

Firefox Developer Edition:
.xls = application/vnd.ms-excel
.xlsx = application/vnd.openxmlformats-officedocument.spreadsheetml.sheet

Safari:
.xls = application/vnd.ms-excel
.xlsx = application/vnd.openxmlformats-officedocument.spreadsheetml.sheet
```

I have Microsoft Excel 365 installed on my Mac, if that makes a difference (cursory research suggests that it does https://github.com/thoughtbot/paperclip/issues/2170, although perhaps more on Windows than Mac).

I'll do a quick test from a Windows 7 laptop as well to record the ones we see there.

[the remark this answers is not in the document]
Comment (rajadain) [reply src/django/api/tests.py]: Just tested from a Windows 7 laptop that does not have Excel installed (it has OpenOffice installed instead).  Both Chrome and Firefox used `application/octet-stream`. Could not test IE11 due to some ancillary browser issues. 

I'm good leaving these as is, or changing them to `application/octet-stream`. In my manual testing it doesn't seem to make a difference. And we're not inspecting the mime type to determine the kind of file (relying on the file extension instead).

**Comment by rajadain** [comment]
Updated instructions to include Excel files in 560464f. This is now ready for another look.

**Comment by jwalgran** [comment]
Taking another look at this now

**Comment by rajadain** [comment]
Thanks for reviewing! Going to squash the fixups and merge once green.

**Pull request merged**